crates/ra_hir_ty/src/expr.rs

```diff
@@ -310,3 +315,87 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<ExprId> {
+    let mut unsafe_expr_ids = vec![];
+    let body = db.body(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Call { callee, .. } => {
+                if infer
+                    .method_resolution(/* id */ *callee)
+                    .map(|func| db.function_data(func).is_unsafe)
+                    .unwrap_or(false)
+                {
+                    unsafe_expr_ids.push(id);
+                }
+            }
+            Expr::MethodCall {/*_receiver, method_name,*/ .. } => {
+                // let receiver_ty = &infer.type_of_expr[*receiver];
+                // receiver_ty
+                if infer
+                    .method_resolution(id)
+                    .map(|func| {
+                        db.function_data(func).is_unsafe
+                    })
+                    .unwrap_or_else(|| {
+                        false
+                    })
+                {
+                    unsafe_expr_ids.push(id);
+                }
+            }
+            Expr::UnaryOp { expr, op: UnaryOp::Deref } => {
+                if let Ty::Apply(ApplicationTy { ctor: TypeCtor::RawPtr(..), .. }) = &infer[*expr] {
+                    unsafe_expr_ids.push(id);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    unsafe_expr_ids
+}
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
+    func: FunctionId,
+    infer: Arc<InferenceResult>,
+    sink: &'a mut DiagnosticSink<'b>,
+}
+
+impl<'a, 'b> UnsafeValidator<'a, 'b> {
+    pub fn new(
+        func: FunctionId,
+        infer: Arc<InferenceResult>,
+        sink: &'a mut DiagnosticSink<'b>,
+    ) -> UnsafeValidator<'a, 'b> {
+        UnsafeValidator { func, infer, sink }
+    }
+
+    pub fn validate_body(&mut self, db: &dyn HirDatabase) {
+        let def = self.func.into();
+        let unsafe_expressions = unsafe_expressions(db, self.infer.as_ref(), def);
+        let func_data = db.function_data(self.func);
+        let unnecessary = func_data.is_unsafe && unsafe_expressions.len() == 0;
```

Comment:
hmm, should we really lint about having `unsafe` on a function without unsafe operations? Especially since we won't detect all unsafe operations yet, so this will have lots of false positives, but even disregarding that, it can be correct for a function to be marked `unsafe` without it having any unsafe operations.

Reply:
I removed this diagnostic.

Reply:
`Vec::set_len` is an example of an unsafe function that doesn't contain any unsafe operations.